### ISSUE TITLE
A general question on TimeSeriesKMeans(TSK)

### ISSUE THREAD

**Issue by Ninimama** [tslearn-team/tslearn#286]
Hello tslearn community,

I was wondering how TSK  works with DTW distance. In Euclidean Distance, we know that the goal is to minimize the sum of squares of distances to centers. Right?

However, how it works in DTW? After we initialize the starting points (centroids) for clustering, we assign each sample "x" to its closest centroid. 

The distance is: **d = DTW(x, centroid)**

So, it doesn't matter whether we use d or d**2. Right? But, what happens in the final result? 
Are clusters having a minimum sum of squares OR sum of absolutes? In the doc, it says:
**inertia_: Sum of distances of samples to their closest cluster center.**

However, in the source code, it says: 
`def _compute_inertia(distances, assignments, squared=True):
    "Derive inertia (average of squared distances) from pre-computed
    distances and assignments.`
================================================================

The reason behind my question is that I am trying to implement Hierarchical clustering using DTW distance, and I am trying to understand whether I should consider "the sum of absolutes" or "the sum of squared of distances" as the cost for "ward" linkage.
It should make a difference in the result. right?
=============================================================

Maybe I am missing something here. It would be great if someone can help me understand this matter or provide me with some references.

Best,
NIma

**Comment by Ninimama** [comment]
I think I found my answer. Since, DBA  is calculated based on minimizing sum of squares, we should use the square in the ward for DTW. However, I still don't understand why inertia_ is the sum of distances (to centers).

**Comment by GillesVandewiele** [comment]
I don't think it matters whether you take the squared or absolute value. When calculating manhattan or euclidean distances, you take the point-wise differences. Taking squares there will give more weight to large point-wise differences.

DTW is already an aggregated distance, so squaring that should have no effect (the most similar pair (which you take for hierarchical clustering) will stay the most similar after squaring the distances).

You could try both and see whether there's an actual difference in the results?

The `_compute_intertia` you mention is used within the clustering module to check for convergence.

**Comment by Ninimama** [comment]
Thanks for the response. The result is different. I can provide my code here. 

**Please let me elaborate a little bit.**

I want to use the HC ward "concept", but with DTW distance instead of euclidean distance. So, the linkage distance between two groups of data points G1 = {ts_1, ts_2, ...., ts_nG1}, with dba_center ts_g1, 
and 
G2 = {ts_1, ts_2, ...., ts_nG2} with dba_center ts_g2.

is:

linkage distance = cost(G1 + G2) - cost(G1) - cost(G2)

**where,**
 
cost (G) = sum(dtw(ts_1, dba_center) + ... + dtw(ts_nG, dba_center)) 
OR
cost (G) = sum(dtw(ts_1, dba_center)^2 + ... + dtw(ts_nG, dba_center)^2)
========================

As you can see, we can consider the dtw distance of each member to the centroid, or the square of such distance. And, the result is different. 
Since the typical formula for "ward" distance (for euclidean distance) as mentioned in this link [http://www.stat.cmu.edu/~cshalizi/350/lectures/08/lecture-08.pdf](url) 
and the way DBA is calculated in tslearn (which is minimizing dtw^2), I am saying IF I want to implement ward for dtw, I should better use dtw^2.

**Comment by Ninimama** [comment]
Thank Gilles for your input. It's good to get some sort of confirmation :)

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — I think this can be closed: it's a usage question that's already covered by the docs.

For `TimeSeriesKMeans(metric="dtw")` the objective minimised by DBA-k-means is the sum of *squared* DTW distances inside each cluster (`tslearn/clustering/kmeans.py`), and `inertia_` reports exactly that — same convention as `sklearn.cluster.KMeans.inertia_` for Euclidean. The DBA update (`tslearn/barycenters/dba.py`) iteratively tightens the barycenter to lower this within-cluster sum-of-squared-DTW; it's a local optimisation (DBA is non-convex over barycenters), so re-running with a different seed gives a different local minimum even with the same `n_clusters`. There's no toggle between sum-of-absolute and sum-of-squared — it's always squared, and the docstring of `inertia_` is the source of truth.

If a usage question is the only thing here, I'd suggest either closing it or moving to the GitHub Discussions / Stack Overflow `#tslearn` tag where usage questions tend to get faster eyeballs. Happy to reopen if I missed a concrete behavioural concern.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against `tslearn/clustering/kmeans.py` and `tslearn/barycenters/dba.py`.)

**Comment by charavelg** [comment]
By the way, dbscan is available in tslearn, see [here](https://tslearn.readthedocs.io/en/stable/gen_modules/clustering/tslearn.clustering.TimeSeriesDBSCAN.html). The initial question was answered, please open a new issue / question / discussion for any further information.
